stream/src/main/scala/org/apache/pekko/stream/javadsl/Flow.scala

```diff
@@ -770,6 +770,47 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
         (s: S, out: Out) => f.apply(s, out).toScala,
         (s: S) => onComplete.apply(s).toScala))
 
+  /**
+   * Transform each stream element with the help of a resource.
+   *
+   * The resource creation function is invoked once when the stream is materialized and the returned resource is passed to
+   * the mapping function for mapping the first element. The mapping function returns a mapped element to emit
+   * downstream. The returned `T` MUST NOT be `null` as it is illegal as stream element - according to the Reactive Streams specification.
+   *
+   * The `close` function is called only once when the upstream or downstream finishes or fails. You can do some clean-up here,
+   * and if the returned value is not empty, it will be emitted to the downstream if available, otherwise the value will be dropped.
+   *
+   * Early completion can be done with combination of the [[takeWhile]] operator.
+   *
+   * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
+   *
+   * You can configure the default dispatcher for this Source by changing the `akka.stream.materializer.blocking-io-dispatcher` or
+   * set it for a given Source by using [[ActorAttributes]].
+   *
+   * '''Emits when''' the mapping function returns an element and downstream is ready to consume it
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * @tparam R the type of the resource
+   * @tparam T the type of the output elements
+   * @param create function that creates the resource
+   * @param f function that transforms the upstream element and the resource to output element
+   * @param close function that closes the resource, optionally outputting a last element
+   * @since 1.1.0
+   */
+  def mapWithResource[R, T](
+      create: java.util.function.Supplier[R],
```

Comment:
I checked Flux, where a `Callable` is used, which can throw exception.
But I think, we did not expect any exception to be throw ,so a Supplier.

Reply:
I don't think we should assume `create` call can never fail. Can you use `Callable` instead?

Reply:
I think a user should handle the maybe exception explicitly, with some kind of `try/retry` otherwise they will wondering , wow, why my stream just failed,ouch, I need somekind of retry.

Reply:
Shouldn't we be using one of the interfaces in https://github.com/apache/incubator-pekko/blob/b0fdac259bd57fdd481483f3fe9a7aec6e1ff38a/actor/src/main/scala/org/apache/pekko/japi/function/Function.scala which are deliberately designed to be used within Pekko because they support throwing exceptions in the `apply`?

Reply:
eg: fs2's `Stream#bracket` where `def bracket[F[_], R](acquire: F[R])(release: R => F[Unit]): Stream[F, R]` the `F` is not limited to `Try` too. so does the `Resource.make`.

And if a use change the checked exception to a runtime exception, the flow will fails too, eg: using `@SneakyThrows` .

Reply:
When we expecting an exception to be throw then that function should be used, otherwise, I a more java native one should be better.
For this operator,  user can using a `RestartFlow` with this or with restart strategy to handle exception more properly, the unexpected one.

A function can throw implies that we *wanting* the exception.

Reply:
I don't think its even a question of want, its rather that if a user ever tried to supply a standard Java lambda that happens to have a `throw` they cannot actually do this because it would fail to compile, i.e. not even a 

```scala
new java.lang.Runnable {
  override def run () {
     ...// do some logic
     if (somethingFatal)
        throws new IllegalArgumentException()
  }
}
```

can compile because `Runnable` doesn't allow exceptions in the `@run` method. This is why those `FunctionalInterface`'s can exist in Pekko.

So unless we can say thats it in **_every single case_** its wrong to throw an exception in the provided lambda, we should use the Pekko provided functional interfaces.

Reply:
So I just did a search and found another case where we have a [similar `create` argument ](https://github.com/apache/incubator-pekko/blob/43bab55616f40c0cf5ece07d8cd07110a3027c7d/stream/src/main/scala/org/apache/pekko/stream/javadsl/Sink.scala#L514) and as you can see it uses the pekko `Creator` functional interface (see https://github.com/apache/incubator-pekko/blob/b0fdac259bd57fdd481483f3fe9a7aec6e1ff38a/actor/src/main/scala/org/apache/pekko/japi/function/Function.scala#L90-L97)

I think its genuinely expected that `create` like methods such as this can possible throw exceptions.

Reply:
> If we expecting exceptions, we will need change every function, then the question is why scala dsl doesn't using these?

What do you mean? As far as I can tell, every single `create` allocator style parameter in pekko-streams uses `org.apache.pekko.japi.function.Creator`. While it is true that `java.util.function.Supplier` is also used in pekko-streams, its only for allocate methods where you create an initial data structure (i.e. https://github.com/apache/incubator-pekko/blob/cf704782010db237bd07f197e0942c9b02ffe23c/stream/src/main/scala/org/apache/pekko/stream/javadsl/Flow.scala#L4360) and for a data structure that makes sense (i.e. creating a `Map` doesn't throw any exceptions).

However as far as I can tell, since we are dealing with `mapWithResource` you can easily supply a function in `create` that can throw an exception even in legitimate scenarios i.e. creating a file so using `java.util.function.Supplier` is actually wrong here

> Ad a daily java developer, I really don't like the japi.function.Functions.

Why? Since it implements `@FunctionalInterface` and its a SAM (single abstract method) if you use lambda syntax its exactly the same or am I missing something here?

Reply:
1. for the unexpected exception, simple: 
![image](https://github.com/apache/incubator-pekko/assets/501740/9da57b45-8664-4676-9f63-c0f1b29470cc)
2. if the current usage of `function.Function*` is totally right, why the `aggregateWithBoundary` use `java.util.function.*` and `mergeSorted` use a java `Comparator`?
3. if the we expect api consistency ,why scaladsl doesn't use these fancy handy `@Exception` methods but the one from stdlib?
4. why java's Stream api doesn't use the `X` methods which can throw？
5. why reactor-core, rxjava and smallrye-mutiny don't use these `throw Exception` methods?

I think we should keep the Api more Java **Native**, not define for the end javadsl user, everytime I see those *function.Function*, I feel wired and very hard to promote pekko/akka to java developers.

A user can live with java function, java stream, reactor-core, rxjava and mutiny, they can live with this javaapi too.

Reply:
@mdedetrich You can add a `@SneakThrow` on your `run` method.

Reply:
> if the current usage of `function.Function*` is totally right, why the `aggregateWithBoundary` use `java.util.function.*` and `mergeSorted` use a java `Comparator`?

I already explained why. Some of these functions create a data structure which legitimately should never throw an exception. Other functions in their `create` method could do anything, in which case you use `Creator`. In short if the supplied lambda in 100% of cases would never throw an exception (i.e. creating a collection) then using `Supplier is fine otherwise use `Creator`. The `allocate` parameter in `aggregateWithBoundary` is for creating a collection, not creating **_anything_**.

>  if the we expect api consistency ,why scaladsl doesn't use these fancy handy `@Exception` methods but the one from stdlib?

Because than that would complicate `javadsl` vs `scaladsl`. The `javadsl` are often just wrappers over `scaladsl` and if both use `org.apache.pekko.stream.javadsl.functional` it makes it easier.

> why java's Stream api doesn't use the `X` methods which can throw？

[It does](https://github.com/apache/incubator-pekko/blob/b0fdac259bd57fdd481483f3fe9a7aec6e1ff38a/stream/src/main/scala/org/apache/pekko/stream/javadsl/RestartSink.scala#L195)

>  why reactor-core, rxjava and smallrye-mutiny don't use these `throw Exception` methods?

I don't know and tbh its kind of not that relevant.

> I think we should keep the Api more Java Native, not define for the end javadsl users, everytime I see those function.Function, I feel wired and very hard to promote pekko/akka to java developers.

This is not a discussion to have in this PR. If we are going to make a big shift from `org.apache.pekko.stream.javadsl.functional` interfaces then it should be discussed in the community and not tagged along a feature PR. Any kind of radical shifts like this has to be agreed on collectively.

Until then, we should be consistent what whats currently in Pekko including adding new features. And in this case the interface to use is `Creator` because the lambda supplied to `create` can be anything.

Reply:
>  Some of these functions create a data structure which legitimately should never throw an exception. 

A user can throw everywhere, and a labmda can closingovey everything.
```java
<T extends Throwable, R> R sneakyThrow(Throwable t) throws T {
        throw (T) t;
    }
```
Can hide without a `lombok`. a new user of pekko stream will wonder what's the heck of `function.Function`. 

And you still can't explain why there is no `java.util.function.CheckedConsumer` or `java.util.function.CheckedSupplier` in java 21.

Reply:
> A user can throw everywhere, and a labmda can closingovey everything.

Yes and then you end up polluting your code in Java with these kinds of wrappers. The better way to solve this problem is to just use a `@FunctionalInterface` that allows throwing of exceptions which is exactly what `org.apache.pekko.japi.function.Creator` is. Java users can just provide a function using standard Java 8 lamda synta without having to use lombok or create any other kind of workarounds, I don't see whats the big issue here.

> And you still not explain why there is no `java.util.function.CheckedConsumer` or `java.util.function.CheckedSupplier` in java 21.

It doesn't matter, maybe they will add it in Java 23 or 25, who knows?

I would strongly suggest if you want more buy in for this PR, just change it to `Creator` and move on. This is not a hill worth dying on, or to put it differently its not worth it to block the PR over this. If you want to start a discussion about using `java.lang.*` FunctionalInterface's generally within Pekko `javadsl` (which is **_NOT_** the current case) then make a separate discussion on it.

There is still time, the Pekko 1.1.x release isn't going to happen for some time.

Reply:
` just change it to Creator and move on. ` why not `java.util.concurrent.Callable`? every java developer knows about it.

Reply:
Yes' we can wait a little and see how other people say, people have different tastes, that's common.

Reply:
> `just change it to Creator and move on.` why not `java.util.concurrent.Callable`? every java developer knows about it.

We are going around in circles here, because thats not how the current Pekko codebase is written,  its inconsistent with everything else and I have already shown why. If you want to fundamentally change how we treat functions in `javadsl`, create a separate discussion and don't bundle it with a feature request.

Reply:
Is there a reason why we couldn't add overloaded mapWithResource methods? One version that uses the Pekko Function interfaces (Creator) and a 2nd one that uses Java Function interfaces (Callable or Supplier).

Reply:
Maybe ,under another package

Reply:
I added this to my local copy of javadsl.Flow.

```
  def mapWithResource[R, T](
      create: java.util.function.Supplier[R],
      f: java.util.function.BiFunction[R, Out, T],
      close: java.util.function.Function[R, Optional[T]]): javadsl.Flow[In, T, Mat] = ???

  def mapWithResource[R, T](
      create: function.Creator[R],
      f: java.util.function.BiFunction[R, Out, T],
      close: java.util.function.Function[R, Optional[T]]): javadsl.Flow[In, T, Mat] = ???
```

This compiled ok. Maybe there are other issues when implementations are added but so far, it looks feasible to use overloaded methods.

Reply:
It will not work with lambda

Reply:
Overloading like this will not work. I also realised that `close` needs to use a `org.apache.pekko.stream.javadsl.functional.Procedure` just likes its done [here](https://github.com/apache/incubator-pekko/blob/6872f9c8b78ae2fababb97cb4600d67d77b4bbd2/stream/src/main/scala/org/apache/pekko/stream/javadsl/Source.scala#L854) in other parts of the code